### PR TITLE
OEL-93: Epic europa search

### PR DESCRIPTION
**Pull request by drishu**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -6,36 +6,36 @@
...
-        "symfony/serializer": "^3.4"
+        "symfony/options-resolver": "^3.4 || ^4.3",
+        "symfony/property-info": "^3.4 || ^4.3",
+        "symfony/serializer": "^3.4 || ^4.3"
     },
     "require-dev": {
-        "openeuropa/code-review": "~1.0.0-beta4",
-        "phpunit/phpunit": "^7.5"
-    },
-    "suggest": {
```
Comment (brummbar) [inline composer.json:22]: Restore the suggest. It's meant for people utilising the library.

File: phpunit.xml.dist [hunk 2]
```diff
@@ -9,9 +9,4 @@
             <directory>./tests</directory>
         </testsuite>
     </testsuites>
-    <filter>
-        <whitelist addUncoveredFilesFromWhitelist="true">
-            <directory>./src</directory>
-        </whitelist>
-    </filter>
```
Comment (brummbar) [inline phpunit.xml.dist:16]: Why the test coverage list was removed?

File: composer.json [hunk 3]
```diff
@@ -6,36 +6,36 @@
...
-        "http-interop/http-factory-guzzle": "PSR-17 HTTP Factories implementation for Guzzle",
-        "php-http/guzzle6-adapter": "PSR-18 HTTP Client implementation for Guzzle"
+        "guzzlehttp/psr7": "^1.8",
+        "http-interop/http-factory-guzzle": "^1.0",
+        "openeuropa/code-review": "2.x-dev",
+        "php-http/guzzle7-adapter": "^1.0",
+        "phpunit/phpunit": "^7.5 || ^8 || ^9"
```
Comment (brummbar) [inline composer.json]: Why do we need this? We don't have Drupal around here, so we can freely use the latest major.

File: src/Client.php [hunk 4]
```diff
@@ -1,133 +1,326 @@
...
-        array $configuration = []
-    ) {
-        $this->httpClient = $httpClient;
-        $this->requestFactory = $requestFactory;
-        $this->streamFactory = $streamFactory;
+        UriFactoryInterface $uriFactory,
+        array $configuration
+    ): void {
+        $container = new Container();
+        $container->share('optionResolver', OptionsResolver::class);
```
Comment (brummbar) [inline src/Client.php]: I don't understand why simple classes like the option resolver have been made services. They have state, and this forced workarounds like resetting the options. This has to be reverted.

File: src/Client.php [hunk 5]
```diff
@@ -1,133 +1,326 @@
...
+                'setRequestFactory' => ['requestFactory'],
+                'setStreamFactory' => ['streamFactory'],
+                'setUriFactory' => ['uriFactory'],
+                'setMultipartStreamBuilder' => ['multipartStreamBuilder'],
+                'setSerializer' => ['serializer'],
+                'setJsonEncoder' => ['jsonEncoder'],
+            ]);
+
+        // Keep a reference to the container.
+        $this->setContainer($container);
```
Comment (brummbar) [inline src/Client.php]: I think this is overkill and wrong in some aspects. Why not use a normal dependency injection, instead of setters?

File: src/Contract/ApiInterface.php [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+     *           'type' => ['integer', 'string'],
+     *           'required' => false,
+     *           'default' => 0,
+     *       ],
+     *   ]
+     *   @endcode
+     *
+     * @see \Symfony\Component\OptionsResolver\OptionsResolver
+     */
+    public function getConfigSchema(): array;
```
Comment (brummbar) [inline src/Contract/ApiInterface.php]: This complicates unnecessarily a simple thing such as the option resolver. There is no need for inventing a new system around it.

File: src/Contract/ClientInterface.php [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+        ?array $languages = null,
+        ?array $query = null,
+        ?string $sortField = null,
+        ?string $sortOrder = null,
+        ?int $pageNumber = null,
+        ?int $pageSize = null,
+        ?string $highlightRegex = null,
+        ?int $highlightLimit = null,
+        ?string $sessionToken = null
+    ): Search;
```
Comment (brummbar) [inline src/Contract/ClientInterface.php:42]: Why this was changed from an array of options to a long, long serie of optional parameters?

File: src/Exception/EuropaSearchApiInvalidParameterValueException.php [hunk 8]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OpenEuropa\EuropaSearchClient\Exception;
+
+/**
+ * Thrown when a parameter has a invalid value.
+ */
+class EuropaSearchApiInvalidParameterValueException extends \InvalidArgumentException
```
Comment (brummbar) [inline src/Exception/EuropaSearchApiInvalidParameterValueException.php]: There's no need to namespace with `EuropaSearchApiInvalid`, the namespace is there already.

File: src/Traits/TokenAwareTrait.php [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        $this->tokenService = $tokenService;
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getRequestHeaders(): array
+    {
+        $token = $this->tokenService->getToken();
```
Comment (brummbar) [inline src/Traits/TokenAwareTrait.php]: Token has an expiration time but we always reuse. Maybe we can account for it.

File: src/Client.php [hunk 10]
```diff
@@ -1,133 +1,326 @@
...
  *
- * @todo Add methods to allow fluent calls: $client->ingestion()->ingestText(...)
- * @todo Rename "_endpoint" options as they represent servers.
+ * This class acts like a proxy for the underlying APIs, such as search,
+ * ingestion, token, facets, etc. The caller should instantiate this class as
+ * the only entry-point.
  */
 class Client implements ClientInterface
 {
+    use ContainerAwareTrait;
```
Comment (brummbar) [inline src/Client.php]: This exposes the container with a public method. Why do we need that?

File: tests/src/Api/DeleteApiTest.php [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+            $this->assertEquals('http://example.com/ingest/delete?apiKey=bananas&database=cucumbers&reference=foo', $request->getUri());
+            $this->inspectAuthorizationHeaders($request);
+        }
+    }
+
+    /**
+     * @see self::testDeleteDocument
+     * @return array
+     */
+    public function providerTestDeleteDocument(): array
```
Comment (brummbar) [inline tests/src/Api/DeleteApiTest.php]: No need for data providers when only one test scenario is provided.

File: tests/src/Api/DeleteApiTest.php [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    {
+        $actualResult = $this->getTestingClient($clientConfig, $responses, [$this, 'inspectRequest'])
+            ->deleteDocument('foo');
+        $this->assertSame($expectedResult, $actualResult);
+    }
+
+    /**
+     * @param RequestInterface $request
+     */
+    public function inspectRequest(RequestInterface $request): void
```
Comment (brummbar) [inline tests/src/Api/DeleteApiTest.php]: Why inspect and not assert?

File: tests/src/Api/FileIngestionApiTest.php [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+            $parts = $this->getMultiParts($request);
+            $this->assertCount(6, $parts);
+            $fileData = file_get_contents(__DIR__ . '/../../fixtures/files/image.png');
+            $this->inspectPart($parts[0], 'application/json', 'metadata', 55, '{"field1":["value1","value2"],"field2":["value3",2345]}');
+            $this->inspectPart($parts[1], 'application/json', 'aclUsers', 20, '["user001","user02"]');
+            $this->inspectPart($parts[2], 'application/json', 'aclNolUsers', 20, '["user003","user04"]');
+            $this->inspectPart($parts[3], 'application/json', 'aclGroups', 23, '["group001","group002"]');
+            $this->inspectPart($parts[4], 'application/json', 'aclNoGroups', 23, '["group003","group004"]');
+            $this->inspectPart($parts[5], 'image/png', 'file', 67, $fileData);
+        }
```
Comment (brummbar) [inline tests/src/Api/FileIngestionApiTest.php]: This for me it's very hard to follow, having to resort to if() to run the correct assertions.

File: tests/src/MockHandler.php [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+declare(strict_types=1);
+
+namespace OpenEuropa\Tests\EuropaSearchClient;
+
+use GuzzleHttp\Handler\MockHandler as GuzzleMockHandler;
+use GuzzleHttp\Promise\PromiseInterface;
+use Psr\Http\Message\RequestInterface;
+
+class MockHandler extends GuzzleMockHandler
```
Comment (brummbar) [inline tests/src/MockHandler.php]: No need to do this, you can use Middleware::history() to be able to read the responses.
In Drupal is a bit harder to inject a middleware, but it has been done [successfully here](https://github.com/openeuropa/media_avportal/blob/8.x-1.x/tests/modules/media_avportal_mock/media_avportal_mock.services.yml#L3).

File: tests/src/Api/TokenApiTest.php [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+/**
+ * @coversDefaultClass \OpenEuropa\EuropaSearchClient\Api\TokenApi
+ */
+class TokenApiTest extends TestCase
+{
+    use ClientTestTrait;
+    use InspectTestRequestTrait;
+
+    /**
+     * @covers ::getToken
```
Comment (brummbar) [inline tests/src/Api/TokenApiTest.php]: Any `@covers` limits the code coverage to verify that the code in that method is covered. By running the code coverage at this moment, most of the code is reported as not covered because of the misuse of the `@covers` tags.

File: tests/src/Api/FacetApiTest.php [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+        [$facetsAsArray, $facetsAsObject] = $this->generateTestingFacetItems(5);
+        $facetValues = [];
+        for ($i = 0; $i < 5; $i++) {
+            $facetValues[] = [
+                'apiVersion' => str_shuffle(md5(serialize($facetValues))),
+                'count' => rand(30, 3000),
+                'rawValue' => str_shuffle(md5(serialize($facetValues))),
+                'value' => str_shuffle(md5(serialize($facetValues))),
+            ];
+        }
```
Comment (brummbar) [inline tests/src/Api/FacetApiTest.php]: What is this for? I guess a leftover.

File: tests/src/Api/SearchApiTest.php [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+    {
+        return [
+            'simple search' => [
+                [
+                    'apiKey' => 'foo',
+                    'database' => 'bar',
+                    'searchApiEndpoint' => 'http://example.com/search',
+                ],
+                [
+                    new Response(200, [], json_encode([
```
Comment (brummbar) [inline tests/src/Api/SearchApiTest.php]: This response is clearly better stored in a fixture json file.

File: tests/src/Traits/FacetTestGeneratorTrait.php [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+                'value' => str_shuffle($toShuffle),
+            ];
+        }
+
+        $facetValuesAsObject = array_map(
+            [$this, 'generateTestingFacetValue'],
+            $facetValuesAsArray
+        );
+
+        return [$facetValuesAsArray, $facetValuesAsObject];
```
Comment (brummbar) [inline tests/src/Traits/FacetTestGeneratorTrait.php]: Just return one of them, and expose a method that does the conversion?

File: tests/src/Traits/InspectTestRequestTrait.php [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+        $this->assertSame('JWT_TOKEN', $request->getHeaderLine('Authorization-propagation'));
+    }
+
+    /**
+     * @param RequestInterface $request
+     */
+    protected function inspectBoundary(RequestInterface $request): void
+    {
+        preg_match('/; boundary="([^"].*)"/', $request->getHeaderLine('Content-Type'), $found);
+        $this->boundary = $found[1];
```
Comment (brummbar) [inline tests/src/Traits/InspectTestRequestTrait.php]: Using a property to store the boundary to run assertions across methods, it stinks. Why not extract the boundary and pass it to the methods that rely on it?

**Review by brummbar — CHANGES_REQUESTED** [review]
The major problem of the PR is that a huge amount of changes in architecture have been done, but not discussed when an analysis was requested.
I personally don't agree with many of the changes. The Api classes were meant to host multiple resources, but now they have been converted to single resource classes. The option resolver scope has been twisted, and a container has been introduced, mixing dependencies between the classes.

The reason I requested an analysis *before* starting the work was to see how we could improve the structure and to avoid this situation.
Can you please explain now the reasoning for all the architecture changes above? We need to discuss this and see from there on.

**Comment by brummbar** [comment]
I forgot to write it before, but also method documentation has been forgotten almost everywhere.

**Comment by brummbar** [comment]
Another note: there is no need to allow to configure each endpoint, it's enough to configure the API uri, the endpoints reside inside that API itself. For example the text and file ingest endpoints will reside under the same API (or controller as ES names them).

File: src/Client.php (hunk 5 above)
Comment (claudiu-cristea) [reply src/Client.php]: Setters injection is a _normal dependency injection_ too. Nothing is making it _special_, see https://en.wikipedia.org/wiki/Dependency_injection#Types_of_dependency_injection. Why one was preferred over the other? Because of the stupid PHPMD, which is limiting the number of method params.

File: src/Contract/ApiInterface.php (hunk 6 above)
Comment (claudiu-cristea) [reply src/Contract/ApiInterface.php]: The option resolver is still used this only abstracts the way APIs are using it. Instead of copying over and over the same code we're just collecting the config from API classes.

File: src/Client.php (hunk 4 above)
Comment (claudiu-cristea) [reply src/Client.php]: Fully agree. Using `->add()` instead of `->share()` is fixing the issue.

File: src/Contract/ClientInterface.php (hunk 7 above)
Comment (claudiu-cristea) [reply src/Contract/ClientInterface.php:42]: Well, the array of options (it was `$parameters` if I recall correctly) was a poor approach. First, it's mixing API parameters (page size, page number, etc) with configurations. Conceptually, configuration is something that is relatively persistent. It doesn't change often, while parameters are part of the API call, they are different from call to call. Also we should avoid, as much as we can, dropping everything in an associative array which acts like a sandbox with arbitrary keys. Instead, we should use dedicated properties, with setters & getters which are bringing semantics to our API classes. Think of the tons of refactorings made from Drupal 7 to 8, where a lot of arrays were converted into classes. Even a Drupal config is no more an arbitrary array but has schema, which limits passing arbitrary values. Or think abour this lib, where we have the Search model, not just an array. Second, we follow the Europa Search API. More or less we're exposing the underlying ES API. I think this is the role of PHP client, to wrap the API in a PHP API. Finally, it's about DX (developer experience). Isn't it easier for a developer, using this client, to quickly discover how he/she needs to use the method by simply revealing in the IDE the signature? No need for extensive documentation, the method speaks for itself.

File: tests/src/Api/DeleteApiTest.php (hunk 11 above)
Comment (claudiu-cristea) [reply tests/src/Api/DeleteApiTest.php]: We've decided to use a data provider in order to allow future extension with other test cases. No big deal, it can be reverted.

File: tests/src/Api/FileIngestionApiTest.php (hunk 13 above)
Comment (claudiu-cristea) [reply tests/src/Api/FileIngestionApiTest.php]: This is because the ingestion makes 2 requests. We have somehow to distinguish.

File: tests/src/MockHandler.php (hunk 14 above)
Comment (claudiu-cristea) [reply tests/src/MockHandler.php]: We will research this. Nice if we can do it w/o this hack. I was on the pressure to assert outgoing requests, NOT the responses. With this we've been caught & fix a lot of errors. So, if we can read from somewhere the HTTP calls it's better.

**Comment by claudiu-cristea** [comment]
> I forgot to write it before, but also method documentation has been forgotten almost everywhere.

Well, we can add "documentation" but would you like something like this?

```php
/**
 * Constructs a new instance.
 * 
 * @param int $pageSize
 *   The page size.
 */
public function __construct(int $pageSize) {...}
```
or

```php
/**
 * Sets the page size.
 *
 * @param int $pageSize
 *   The page size.
 */
public function setPageSize(int $pageSize): self {...}
```

I really don't understand why someone would need such a "documentation". For me is only _noise_ and creates the opposite effect, it makes the code harder to read. IMHO, a good documentation is already there if:

* You do properly namings. A good name already tells what's the purpose of that var, property, key, class, method, etc.
* You provide full strict typing. Having a good name, strict typing _documents_ better what is the nature of that thing.

I know the Drupal way of putting useless "docs" everywhere, but I disagree with this dogma. As this is not Drupal code, I've assumed that we can relax this in the same way as you can see in a lot of popular PHP repos. But...

> Another note: there is no need to allow to configure each endpoint, it's enough to configure the API uri, the endpoints reside inside that API itself. For example the text and file ingest endpoints will reside under the same API (or controller as ES names them).

I see this point and I can resonate with the idea. It was on our mind from the beginning. However, long term is a better approach. In the high-performance database practices the read and write are often deferred to different backends (see the master-slave replication). At some point, ES might take a similar approach. Also, we initially thought to store the base URL and only the relative paths of each API. But then it's the same number of configs plus 1. Hardcoding the paths? I think this was initially done but I see the ES team is actively working and changing their API (some additions will follow our suggestions), so we thought to allow the full flexibility.

File: src/Contract/ClientInterface.php (hunk 7 above)
Comment (brummbar) [reply src/Contract/ClientInterface.php:42]: Yes, the configuration of the API should not be passed along with the endpoint parameters. But the options resolver was there to validate the possible options, as they are quite a big number. Now the method has dozen of parameters, and to pass only two of them you have to `search(null, null, null, null, null, $value, null, null, null, $value)`.
I would have kept required parameters as dedicated ones, and the rest as an array. Or a value object, and passing that in the method itself. Furthermore, the Api classes have all these setters, but they are not really exposed ever to the user, as the readme states that only the client should be used to call them.

File: composer.json (hunk 1 above)
Comment (claudiu-cristea) [reply composer.json:22]: Add a Guzzle 7 latest or PSR compliant

File: phpunit.xml.dist (hunk 2 above)
Comment (claudiu-cristea) [reply phpunit.xml.dist:16]: We add the removed hunk back

File: src/Client.php (hunk 5 above)
Comment (claudiu-cristea) [reply src/Client.php]: Turn off PHPMD (grumphp.yml) and revert back to constructor injection.

File: src/Contract/ApiInterface.php (hunk 6 above)
Comment (claudiu-cristea) [reply src/Contract/ApiInterface.php]: Client should pass only the config intended for a particular API class. Move config validation from setter to constructor.

File: tests/src/Api/FacetApiTest.php (hunk 16 above)
Comment (claudiu-cristea) [reply tests/src/Api/FacetApiTest.php]: Will look into...

**Review by drishu — COMMENTED** [review]
Regarding the comments (method, classes etc) we make sure to comment just the ones that are not 100% clear.
We also discussed naming the methods for executing the search all the same way as the API classed are Endpoint classes.
Lets to 1 endpoint changed to see the impact and discuss.

File: tests/src/Api/DeleteApiTest.php (hunk 11 above)
Comment (drishu) [reply tests/src/Api/DeleteApiTest.php]: We wil extend testing in
https://citnet.tech.ec.europa.eu/CITnet/jira/browse/OEL-296
not high prio right now

File: tests/src/Api/FacetApiTest.php (hunk 16 above)
Comment (timpevi) [reply tests/src/Api/FacetApiTest.php]: It comes from this https://github.com/openeuropa/europa-search-client/commit/cb1b513c2091dec1b15fa120583829039f46102a

it can be removed.

File: composer.json (hunk 1 above)
Comment (drishu) [reply composer.json:22]: done

File: phpunit.xml.dist (hunk 2 above)
Comment (drishu) [reply phpunit.xml.dist:16]: done

File: composer.json (hunk 3 above)
Comment (drishu) [reply composer.json]: done

File: src/Client.php (hunk 5 above)
Comment (drishu) [reply src/Client.php]: phpmd turned off

File: src/Exception/EuropaSearchApiInvalidParameterValueException.php (hunk 8 above)
Comment (drishu) [reply src/Exception/EuropaSearchApiInvalidParameterValueException.php]: done

File: src/Client.php (hunk 10 above)
Comment (drishu) [reply src/Client.php]: To expose it in case 3rd party wants to inject their own stuff, but we removed it and well add back in case the need arises

File: tests/src/Api/DeleteApiTest.php (hunk 12 above)
Comment (drishu) [reply tests/src/Api/DeleteApiTest.php]: done

File: tests/src/Api/FileIngestionApiTest.php (hunk 13 above)
Comment (drishu) [reply tests/src/Api/FileIngestionApiTest.php]: fixed with history middleware

File: tests/src/MockHandler.php (hunk 14 above)
Comment (drishu) [reply tests/src/MockHandler.php]: done

File: tests/src/Api/TokenApiTest.php (hunk 15 above)
Comment (drishu) [reply tests/src/Api/TokenApiTest.php]: done

File: tests/src/Api/FacetApiTest.php (hunk 16 above)
Comment (drishu) [reply tests/src/Api/FacetApiTest.php]: removed

File: tests/src/Api/SearchApiTest.php (hunk 17 above)
Comment (drishu) [reply tests/src/Api/SearchApiTest.php]: done

File: tests/src/Traits/FacetTestGeneratorTrait.php (hunk 18 above)
Comment (drishu) [reply tests/src/Traits/FacetTestGeneratorTrait.php]: done

File: tests/src/Traits/InspectTestRequestTrait.php (hunk 19 above)
Comment (drishu) [reply tests/src/Traits/InspectTestRequestTrait.php]: done

File: src/Client.php (hunk 4 above)
Comment (drishu) [reply src/Client.php]: for now we did it like @claudiu-cristea recommended, moving them back into the api classes handled in POC here https://github.com/openeuropa/europa-search-client/pull/58

File: src/Client.php (hunk 5 above)
Comment (drishu) [reply src/Client.php]: the setters for utility classes handled in POC https://github.com/openeuropa/europa-search-client/pull/58

File: src/Contract/ApiInterface.php (hunk 6 above)
Comment (drishu) [reply src/Contract/ApiInterface.php]: handled in POC https://github.com/openeuropa/europa-search-client/pull/58

File: src/Contract/ClientInterface.php (hunk 7 above)
Comment (drishu) [reply src/Contract/ClientInterface.php:42]: in this case optionsResolver totally makes sense as a service, it has state, would then be needed in different scenarios/classes.. lets discuss

File: src/Traits/TokenAwareTrait.php (hunk 9 above)
Comment (drishu) [reply src/Traits/TokenAwareTrait.php]: We mentioned this to Stefan Kolev prior, he told us that its a cheap operation, that you get a new token only if the old one expires for any subsequent call, and that the current java client also did not implement it.
Agreed that we can save some http calls, we've created OEL-370 for this

**Review by brummbar — APPROVED** [review]
